src/requests/is-verified-request.ts

```diff
@@ -0,0 +1,84 @@
+import {
+  CanonicalRequest,
+  CanonicalRequestValidator,
+  RequestMetadata,
+  RequestMetadataValidator,
+  Secret,
+  SecretValidator,
+  Timestamp,
+} from './typings'
+import { ContentfulSigningHeader, CONTENTFUL_SIGNING_HEADERS } from './constants'
+import { getNormalizedHeaders, pickHeaders } from './utils'
+import { createSignature } from './create-signature'
+
+const getRequestMetadata = (canonicalRequest: CanonicalRequest): RequestMetadata => {
+  const normalizedHeaders = getNormalizedHeaders(canonicalRequest.headers ?? {})
+
+  const signingHeaders = normalizedHeaders.filter(([key]) =>
+    CONTENTFUL_SIGNING_HEADERS.includes(key)
+  )
+
+  // In order to not rely in order we perform a find. Array is short, not a big deal
+  const [, signature] =
+    signingHeaders.find(([key]) => key === ContentfulSigningHeader.Signature) ?? []
+  const [, rawSignedHeaders] =
+    signingHeaders.find(([key]) => key === ContentfulSigningHeader.SignedHeaders) ?? []
+  const [, rawTimestamp] =
+    signingHeaders.find(([key]) => key === ContentfulSigningHeader.Timestamp) ?? []
+
+  const signedHeaders = rawSignedHeaders ? rawSignedHeaders.split(',') : []
+  const timestamp = Number.parseInt(rawTimestamp ?? '', 10)
+
+  return RequestMetadataValidator.check({ signature, signedHeaders, timestamp })
+}
+
+const isRequestTimestampTooOld = (ttl: number, timestamp: Timestamp) => {
+  return Date.now() - timestamp >= ttl * 1000
+}
+
+/**
+ * Given a secret verifies a CanonicalRequest
+ *
+ * ~~~
+ * const {isVerifiedRequest} = require('contentful-node-apps-toolkit')
+ * const {server} = require('./imaginary-server')
+ * const {makeCanonicalRequestFromImaginaryServerRequest} = require('./imaginary-utils')
+ *
+ * const SECRET = process.env.SECRET
+ * const REQUEST_TTL = Number.parseInt(process.env.REQUEST_TTL, 10)
+ *
+ * server.post('/api/my-resources', (req, res) => {
+ *   const canonicalRequest = makeCanonicalRequestFromImaginaryServerRequest(req)
+ *
+ *   if (!isVerifiedRequest(SECRET, canonicalRequest, REQUEST_TTL)) {
+ *     res.send(403, 'Invalid signature')
+ *   }
+ *
+ *   // Rest of the code
+ * })
+ *
+ * ~~~
+ */
+export const isVerifiedRequest = (
+  rawSecret: Secret,
+  rawCanonicalRequest: CanonicalRequest,
+  ttl: number = 30
+): boolean => {
+  const canonicalRequest = CanonicalRequestValidator.check(rawCanonicalRequest)
+  const secret = SecretValidator.check(rawSecret)
+
+  const { signature, signedHeaders, timestamp } = getRequestMetadata(canonicalRequest)
+
+  if (isRequestTimestampTooOld(ttl, timestamp)) {
+    return false
```

Comment:
I have mixed feelings about this. It kinda feels borderline between an exception (as we do for other validation) and something ordinary we just want to return a `false` for. WDYT?

Reply:
Especially if the user uses a custom TTL I think it would be nice to expose the failure reason, so they can better debug/log their settings and access. Between through throwing errors of different types, or adding a second return parameter, I'm leaning more towards the former

Reply:
Why would you prefer a second return type? Throwing specific errors feels quite clean to me, as long we make it clear what the possibilities are.

Reply:
Maybe we should return something like Either/Maybe type?

Reply:
I wouldn't force consumers of this library to use fp-ts too

Reply:
wdyt of implementing this with a flag? i.e: `ttl` is optional? If it exists - we can throw/return false but if someone doesn't care about it - can still ignore it?

Reply:
I think @Jwhiles was referring more to the concept than the library itself. It's a perfect fit for this case, however not everyone is familiar with functional programming, so I probably wouldn't do it anyway, but it looks to me that the consensus is more around returning a value than throwing an exception.

If that's the case, I would probably leave it as is and add a log line. WDYT?

Reply:
I wouldn't want to pollute people's logs or circumvent their private logging solutions. Throwing an error still seems like a good solution to me.

```js
// App backend request handler
try {
  await verifyRequest(req)
} catch (e) {
  // e.message one of ["Invalid signature", "Missing signature", "Expired timestamp"]
}
```

Reply:
Sorry @kdamball , haven't seen your message before.

> wdyt of implementing this with a flag? i.e: ttl is optional? If it exists - we can throw/return false but if someone doesn't care about it - can still ignore it?

Yeah, I also think that the TTL is too much of an opinionated choice and not everyone necessarily cares, as it's probably part of business logic of your own app

Reply:
A default TTL of 30 seconds shouldn't get in anyone's way, except if they are forwarding it through like ~40 Lambda functions? It also feels like one of those "start strict and open up later" situations